File: solutions/binsearch/binsearch.cpp
#include <iostream>
#include <vector>

int search(std::vector<int> nums, int q, int start, int end) {
    if (end-start <= 2) {
        if (nums[start] == q) return start;
        else return start+1;
    }

    int mid = (end+start) / 2;
    if (nums[mid] > q) return search(nums, q, start, mid);
    else return search(nums, q, mid, end);
}

int main() {
    std::vector<int> nums, queries;
    int len, num_q;

    std::cin >> len >> num_q;
    for (auto i = 0; i < len; i++) {
        int tmp;
        std::cin >> tmp;
        nums.push_back(tmp);
    }
    for (auto i = 0; i < num_q; i++) {
        int tmp;
        std::cin >> tmp;
        queries.push_back(tmp);
    }

    for (auto q: queries) {
        std::cout << search(nums, q, 0, len) << std::endl;
    }
}

File: solutions/binsearch/binsearch.py
def search(nums, q, start=0, end=None):
    if end is None:
        end = len(nums)

    if end - start <= 2:
        return start if nums[start] == q else start+1

    mid = int((end+start) / 2)
    if nums[mid] > q:
        return search(nums, q, start, mid)
    else:
        return search(nums, q, mid, end)

input()
nums = list(map(int, input().strip().split()))
queries = list(map(int, input().strip().split()))
for q in queries:
    print(search(nums, q))
